operator/pkg/controllers/reconcilers/hooks.go

```diff
@@ -68,22 +68,26 @@ func (r *HookReconciler) Reconcile(ctx context.Context, bkapp *paasv1alpha2.BkAp
 		}
 
 		switch {
-		case current.Timeout(resources.HookExecuteTimeoutThreshold):
+		case current.ProgressingWithTimeout(resources.HookExecuteTimeoutThreshold):
 			// 删除超时的 pod
 			if err := r.Client.Delete(ctx, current.Pod); err != nil {
 				return r.Result.WithError(errors.WithStack(resources.ErrExecuteTimeout))
 			}
 			return r.Result.WithError(errors.WithStack(resources.ErrExecuteTimeout))
-		case current.Progressing():
-			return r.Result.requeue(paasv1alpha2.DefaultRequeueAfter)
-		case current.Succeeded():
-			return r.Result
-		case current.FailedUntilTimeout(resources.HookExecuteFailedTimeoutThreshold):
+		case current.FailedWithTimeout(resources.HookExecuteFailedTimeoutThreshold):
 			if err := r.Client.Delete(ctx, current.Pod); err != nil {
 				return r.Result.WithError(errors.WithStack(resources.ErrPodEndsUnsuccessfully))
 			}
 			// Pod 在超时时间内一直失败, 终止调和循环
-			return r.Result.WithError(errors.WithStack(resources.ErrPodEndsUnsuccessfully)).End()
+			log.Error(errors.WithStack(resources.ErrPodEndsUnsuccessfully), "execute timeout")
+			// 不能调用 WithError 否则不会停止调和循环(Result 协议保持与 controller-runtime 的协议一致)
+			// ref: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/reconcile#Reconciler
+			return r.Result.End()
+		case current.Progressing():
+			// 当 Hook 执行成功时会有另外的事件触发新的调和循环, 因此 Hook 执行中的轮询间隔无需太频繁
```

Comment:
Nit：

- 建议注释明确是由什么事件触发新循环，是 owned Pod 逻辑吗？
- 考虑保留原有的 enqueue delay 或者直接删除 requeue 逻辑？现在这样的 delay 太长并且 * 10 看上去没有含义（为什么不是 * 100？），实际也起不到效果，具体效果跟删除差不多，不如删了更明确。

Reply:
- 对，是 owned Pod
- 如果要去掉 requeue 的话，还需要调整下 owned pod 的谓词，让他在 status 变成失败是也触发循环。


其实我更倾向去掉 requeue，只是去掉的话需要测试更多边界，这个 PR 是观测到调和太频繁这个问题顺手提的，所以就不是很想这次处理。

Reply:
那要不这次别修改成 *10？不扩大间隔，代码稍微更好理解一点。

Reply:
一步到位吧, 明确这里 requeue 的意图是处理超时事件，然后外面 owned pod 就处理 hook 执行成功/失败的事件。

Reply:
我觉得白名单的方式会让意图更明确，能更清晰表明什么事件能触发调和循环。(意图明确可以避免调和循环一致被触发，但是不知道究竟是什么事件导致的）


另外社区提供的内置实现都是白名单，注释里也有写清楚使用场景。

- GenerationChangedPredicate: With this predicate, any update events with writes only to the status field will not be reconciled.
- AnnotationChangedPredicate:  This is mostly useful for controllers that needs to trigger both when the resource's generation is incremented(i.e., when the resource' .spec changes), or an annotation changes (e.g., for a staging/alpha API).
- LabelChangedPredicate:  This will be helpful when object's labels is carrying some extra specification information beyond object's spec, and the controller will be triggered if any valid spec change (not only in spec, but also in labels) happens.

Reply:
好的，没问题。